### PR TITLE
Document Native Code Generations

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Documents: #762 
Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: docs/NativeCodeGeneration.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+The first step of the process is identical to the traditional build pipeline. The eBPF program is provided as C or other high level language file and the normal LLVM / Clang tool chain is invoked, to produce a ELF file containing one or more program and zero or more map definitions.
+
+## Step 2 – Verification of the eBPF program
+
+The second step is to use the Prevail verifier to determine if the resulting eBPF program conforms to the required safety requirements. This is identical to the process used to by eBPF for Windows today. Only once the Prevail verifier can prove the program is safe does the process proceed to the next step.
+
+## Step 3 – Generation of the C code from the eBPF byte code
+
+The third step is to generate C code that maps one to one with the eBPF instructions in the ELF file. The eBPF registers are mapped to local variables in the generated C function and each eBPF instruction is used to generate equivalent C operations. In addition, the generated C file also contains bindings for the any helper functions referenced by the eBPF program and any maps the eBPF program uses.
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:27]:
```suggestion
The third step is to generate C code that maps one to one with the eBPF instructions in the ELF file. The eBPF registers are mapped to local variables in the generated C function and each eBPF instruction is used to generate equivalent C operations. In addition, the generated C file also contains bindings for any helper functions referenced by the eBPF program and any maps the eBPF program uses.
```

File: docs/NativeCodeGeneration.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+## Step 4 – Compile generate C code into native COFF file
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:29]:
```suggestion
## Step 4 – Compile generated C code into native COFF file
```

File: docs/NativeCodeGeneration.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+The fourth step is to use the standard C compiler (GCC, MSVC, or Clang) to generate a x64/ARM64 COFF file with Windows calling conventions. The C file will expose a single top-level global variable that allows the skeleton code to find the helpers, maps, and programs contained in the file.
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:31]:
```suggestion
The fourth step is to use a standard C compiler (GCC, MSVC, or Clang) to generate an x64/ARM64 COFF file with Windows calling conventions. The C file will expose a single top-level global variable that allows the skeleton code to find the helpers, maps, and programs contained in the file.
```

File: docs/NativeCodeGeneration.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+The fifth is to use the platform-specific linker to generate the final PE file. The linker consumed the object file from step 4, a generic DriverEntry skeleton.
+
+## Step 6 – Signing the PE file
+
+The last step is to use the standard PE signing tools to produce a signed PE file from the output of step 5. For the developer workflow the signature would be based on a test signing root, whereas for production use the PE file would be signed using the Microsoft PRS signing keys.
+
+# Stub files consumed by the linker
+
+Step 5 consumes a set of object files to generate the final PE image. The following section gives a of the driver skeleton.
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:43]:
```suggestion
Step 5 consumes a set of object files to generate the final PE image. The following section gives a summary of the driver skeleton.
```

File: docs/NativeCodeGeneration.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+## DriverEntry skeleton
+
+All Windows Drivers need to export a static function labelled &quot;DriverEntry&quot; that is invoked by the OS (Operating System) during loading of the driver. In addition to the standard driver related functionality, the driver entry skeleton would also register three NMR contracts exposed by the other framework components (to export the program to the eBPF runtime, to create any maps required, and to import the helper functions).
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:47]:
```suggestion
All Windows Drivers need to export a static function labelled &quot;DriverEntry&quot; that is invoked by the OS (Operating System) during loading of the driver. In addition to the standard driver related functionality, the driver entry skeleton would also register three [Network Module Registrar (NMR)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/introduction-to-the-network-module-registrar) contracts exposed by the other framework components (to export the program to the eBPF runtime, to create any maps required, and to import the helper functions).
```

File: docs/NativeCodeGeneration.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+## Global entry point for generated code
+
+Every generated C file will contain a single global entry point of type meta_data_table_t:
+
+```
+typedef struct _meta_data_table
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:54]:
```suggestion
typedef struct _metadata_table
```
metadata is one word.  replace "meta_data" and "meta-data" with "metadata" throughout this doc.

File: docs/NativeCodeGeneration.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+static map_entry_t_maps[] = {
+    { NULL, { 36, 2, 4, 8, 1,  }, "dropped_packet_map" },
+    { NULL, { 36, 2, 4, 4, 1,  }, "interface_index_map" },
+};
+```
+
+The skeleton then uses NMR to either create an instance of the map or obtain the address of an existing map and then write the address into the address field. References to the maps in the generated code are then indirect via the address field.
+
+## Loading an EBPF program from a PE .sys file
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:121]:
```suggestion
## Loading an eBPF program from a PE .sys file
```

File: docs/NativeCodeGeneration.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+The process of loading an eBPF program is a series of interactions between the eBPF Execution Context and the generated .sys file.
+
+### Lifecycle of a generaed PE .sys file
+
+1) During driver entry, the generated DriverEntry will call [NmrRegisterClient](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/nf-netioddk-nmrregisterclient) for the helper contract and the map contract.
+
+2) In response to the attach notification on the helper function NMR contract, the skeleton will query the address of each helper function the eBPF program used and write the address into the _helpers table.
+
+3) In response to the attach notification on the map resolve NMR contract, the skeleton call into the eBPF execution context to create the maps based on the definitions and write the addresses into the _maps table.
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:131]:
```suggestion
3) In response to the attach notification on the map resolve NMR contract, the skeleton will call into the eBPF execution context to create the maps based on the definitions and write the addresses into the _maps table.
```

File: docs/NativeCodeGeneration.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+## Loading an EBPF program from a PE .sys file
+
+The process of loading an eBPF program is a series of interactions between the eBPF Execution Context and the generated .sys file.
+
+### Lifecycle of a generaed PE .sys file
+
+1) During driver entry, the generated DriverEntry will call [NmrRegisterClient](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/nf-netioddk-nmrregisterclient) for the helper contract and the map contract.
+
+2) In response to the attach notification on the helper function NMR contract, the skeleton will query the address of each helper function the eBPF program used and write the address into the _helpers table.
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:129]: Do you have to do anything special to make tail calls work?  (I don't think so, but want to know if you've verified that.)

Comment (Alan-Jowett) [reply docs/NativeCodeGeneration.md]: This is part of the helper function metadata that is returned from the EC. I added clarifying text.

File: docs/NativeCodeGeneration.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+The skeleton then uses NMR to either create an instance of the map or obtain the address of an existing map and then
+write the address into the address field. References to the maps in the generated code are then indirect via the
+address field.
+
+## Loading an eBPF program from a PE .sys file
+
+The process of loading an eBPF program is a series of interactions between the eBPF Execution Context and the generated
+.sys file.
+
+### Lifecycle of a generaed PE .sys file
```
Comment (dthaler) [inline docs/NativeCodeGeneration.md:163]:
```suggestion
### Lifecycle of a generated PE .sys file
```

**Review by dthaler — APPROVED** [review]

File: docs/NativeCodeGeneration.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+## Exported programs
+
+Each program in the generated C file is exported via a program_entry_t:
+```
+typedef struct _program_entry
+{
+    uint64_t (*function)(void*);
+    const char* name;
+} program_entry_t;
```
Comment (saxena-anurag) [inline docs/NativeCodeGeneration.md:107]: Just mentioning that this will require revision as this is already deviating from the code.

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**